File: src/MicroService.ApiGateway.Domain/Entites/Ocelot/ReRoute.cs
using Volo.Abp.Domain.Entities;

namespace MicroService.ApiGateway.Entites.Ocelot
{
    public class ReRoute : AggregateRoot<int>
    {
        /// <summary>
        /// 路由ID
        /// </summary>
        public virtual long ReRouteId { get; protected set; }
        /// <summary>
        /// 路由名称
        /// </summary>
        public virtual string ReRouteName { get; protected set; }
        /// <summary>
        /// 下游路由路径
        /// </summary>
        public virtual string DownstreamPathTemplate { get; protected set; }
        /// <summary>
        /// 上游路由路径
        /// </summary>
        public virtual string UpstreamPathTemplate { get; protected set; }
        /// <summary>
        /// 下游Http方法列表,分号间隔
        /// </summary>
        public virtual string UpstreamHttpMethod { get; protected set; }
        /// <summary>
        /// 下游请求头
        /// </summary>
        public virtual string AddHeadersToRequest { get; protected set; }
        /// <summary>
        /// 上游请求头变更
        /// </summary>
        public virtual string UpstreamHeaderTransform { get; protected set; }
        /// <summary>
        /// 下游请求头变更
        /// </summary>
        public virtual string DownstreamHeaderTransform { get; protected set; }
        /// <summary>
        /// 添加请求令牌
        /// </summary>
        public virtual string AddClaimsToRequest { get; protected set; }
        public virtual string RouteClaimsRequirement { get; protected set; }
        /// <summary>
        /// 添加请求参数
        /// </summary>
        public virtual string AddQueriesToRequest { get; protected set; }
        public virtual string RequestIdKey { get; set; }
        public virtual CacheOptions CacheOptions { get; protected set; }
        public virtual bool ReRouteIsCaseSensitive { get; set; }
        public virtual string ServiceName { get; set; }
        public virtual string DownstreamScheme { get; set; }
        public virtual QoSOptions QoSOptions { get; protected set; }
        public virtual LoadBalancerOptions LoadBalancerOptions { get; protected set; }
        public virtual RateLimitRule RateLimitOptions { get; protected set; }
        public virtual AuthenticationOptions AuthenticationOptions { get; protected set; }
        public virtual HttpHandlerOptions HttpHandlerOptions { get; protected set; }
        public virtual string DownstreamHostAndPorts { get; protected set; }
        public virtual string DelegatingHandlers { get; set; }
        public virtual string UpstreamHost { get; set; }
        public virtual string Key { get; set; }
        public virtual int? Priority { get; set; }
        public virtual int? Timeout { get; set; }
        public virtual bool DangerousAcceptAnyServerCertificateValidator { get; set; }
        public virtual SecurityOptions SecurityOptions { get; protected set; }

        protected ReRoute()
        {

        }

        public ReRoute(long rerouteId, string routeName, string downPath, string upPath, string upMethod)
        {
            ReRouteId = rerouteId;
            ModifyRouteInfo(routeName, downPath, upPath, upMethod);
            InitlizaReRoute();
        }

        public void ModifyRouteInfo(string routeName, string downPath, string upPath, string upMethod)
        {
            ReRouteName = routeName;
            DownstreamPathTemplate = downPath;
            UpstreamPathTemplate = upPath;
            UpstreamHttpMethod = upMethod;
        }

        public void AddRequestHeader(Headers headers)
        {
            AddHeadersToRequest += $"{headers.Key}:{headers.Value};";
        }

        public void AddUpStreamHeader(Headers headers)
        {
            UpstreamHeaderTransform += $"{headers.Key}:{headers.Value};";
        }

        public void AddDownStreamHeader(Headers headers)
        {
            DownstreamHeaderTransform += $"{headers.Key}:{headers.Value};";
        }

        public void AddRequestClaim(Headers headers)
        {
            AddClaimsToRequest += $"{headers.Key}:{headers.Value};";
        }

        public void AddRequirementCalim(Headers headers)
        {
            RouteClaimsRequirement += $"{headers.Key}:{headers.Value};";
        }

        public void AddRequestQueries(Headers headers)
        {
            AddQueriesToRequest += $"{headers.Key}:{headers.Value};";
        }

        public void AddDownStreamHostAndPort(HostAndPort hostAndPort)
        {
            DownstreamHostAndPorts += $"{hostAndPort.Host}:{hostAndPort.Port};";
        }
        
        private void InitlizaReRoute()
        {
            QoSOptions = new QoSOptions(ReRouteId);
            CacheOptions = new CacheOptions(ReRouteId);
            LoadBalancerOptions = new LoadBalancerOptions(ReRouteId);
            RateLimitOptions = new RateLimitRule(ReRouteId);
            AuthenticationOptions = new AuthenticationOptions(ReRouteId);
            HttpHandlerOptions = new HttpHandlerOptions(ReRouteId);
            SecurityOptions = new SecurityOptions(ReRouteId);
        }
    }
}

File: src/MicroService.ApiGateway.Web/wwwroot/js/ocelot/reroute.js
layui.use(['form', 'element'], function () {
    var form = layui.form,
        element = layui.element,
        $ = layui.jquery;

    var _reRouteRoot= microService.apiGateway.ocelot.reRoute;

    var _reRouteLoadService = _reRouteRoot.get;
    let _reRouteEditService;

    form.on('submit(btnSave)', function (data) {
        _reRouteEditService = $('#ReRouteId').val() == 0
            ? _reRouteRoot.create
            : _reRouteRoot.update;
        let submitDto = {
            reRouteId: $('#ReRouteId').val(),
            reRouteName: data.field.ReRouteName,
            downstreamPathTemplate: data.field.DownstreamPathTemplate,
            upstreamPathTemplate: data.field.UpstreamPathTemplate,
            upstreamHttpMethod: data.field.UpstreamHttpMethod,
            addHeadersToRequest: data.field.AddHeadersToRequest,
            upstreamHeaderTransform: data.field.UpstreamHeaderTransform,
            downstreamHeaderTransform: data.field.DownstreamHeaderTransform,
            downstreamHostAndPorts: data.field.DownstreamHostAndPorts,
            addClaimsToRequest: data.field.AddClaimsToRequest,
            routeClaimsRequirement: "",
            addQueriesToRequest: data.field.AddQueriesToRequest,
            requestIdKey: data.field.RequestIdKey,
            reRouteIsCaseSensitive: true,
            serviceName: data.field.ServiceName,
            downstreamScheme: data.field.DownstreamScheme,
            httpHandlerOptions: {
                allowAutoRedirect: true,
                useCookieContainer: true,
                useTracing: true,
                useProxy: true
            },
            loadBalancerOptions: {
                type: data.field.Type
            },
            rateLimitOptions: {
                clientWhitelist: data.field.ClientWhitelist.split(","),
                enableRateLimiting: $("input[name='EnableRateLimiting']:checked").val() === "on",
                limit: data.field.Limit,
                period: data.field.Period,
                periodTimespan: data.field.PeriodTimespan
            },
            qoSOptions: {
                exceptionsAllowedBeforeBreaking: data.field.ExceptionsAllowedBeforeBreaking,
                durationOfBreak: data.field.DurationOfBreak,
                timeoutValue: data.field.TimeoutValue
            },
            securityOptions: {
                ipAllowedList: data.field.IpAllowedList.split(","),
                ipBlockedList: data.field.IpBlockedList.split(",")
            },
            authenticationOptions: {
                allowedScopes: data.field.AllowedScopes.split(","),
                authenticationProviderKey: data.field.AuthenticationProviderKey
            },
            dangerousAcceptAnyServerCertificateValidator: $("input[name='CertificateValidator']:checked").val() === "on"
        };

        _reRouteEditService(submitDto).done(function (result) {
            console.log(result);
            abp.notify.success('保存成功!');
        });
        return false;
    });

    function refreshTagsInput() {
        $('.tags-input').tagsinput();
    }

    $(function(){
        _reRouteLoadService($('#ReRouteId').val())
            .done(function (result) {
                form.val('ReRouteCfg', {
                    //basic
                    'ReRouteId': result.reRouteId,
                    'ReRouteName': result.reRouteName,
                    'ServiceName': result.serviceName,
                    'UpstreamPathTemplate': result.upstreamPathTemplate,
                    'DownstreamPathTemplate': result.downstreamPathTemplate,
                    'UpstreamHttpMethod': result.upstreamHttpMethod,
                    'DownstreamHostAndPorts': result.downstreamHostAndPorts,
                    'DownstreamScheme': result.downstreamScheme,
                    'RequestIdKey': result.requestIdKey,
                    'CertificateValidator': result.dangerousAcceptAnyServerCertificateValidator,
                    //http
                    'UpstreamHeaderTransform': result.upstreamHeaderTransform,
                    'DownstreamHeaderTransform': result.downstreamHeaderTransform,
                    'AddHeadersToRequest': result.addHeadersToRequest,
                    'AddClaimsToRequest': result.addClaimsToRequest,
                    'AddQueriesToRequest': result.addQueriesToRequest,
                    //ratelimit
                    'ClientWhitelist': result.rateLimitOptions.clientWhitelist.join(','),
                    'EnableRateLimiting': result.rateLimitOptions.enableRateLimiting,
                    'Period': result.rateLimitOptions.period,
                    'PeriodTimespan': result.rateLimitOptions.periodTimespan,
                    'Limit': result.rateLimitOptions.limit,
                    //qos
                    'ExceptionsAllowedBeforeBreaking': result.qoSOptions.exceptionsAllowedBeforeBreaking,
                    'DurationOfBreak': result.qoSOptions.durationOfBreak,
                    'TimeoutValue': result.qoSOptions.timeoutValue,
                    'Type': result.loadBalancerOptions.type,
                    //security
                    'AuthenticationProviderKey': result.authenticationOptions.authenticationProviderKey,
                    'AllowedScopes': result.authenticationOptions.allowedScopes.join(','),
                    'IpAllowedList': result.securityOptions.ipAllowedList.join(','),
                    'IpBlockedList': result.securityOptions.ipBlockedList.join(',')
                });
                refreshTagsInput();
            });
    });
});